go/libraries/doltcore/table/editor/keyless_table_editor.go

```diff
@@ -17,11 +17,358 @@ package editor
+		return nil, err
+	}
+
+	err = kte.eg.Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	return kte.tbl, nil
+}
+
+// Schema implements TableEditor.
+func (kte *keylessTableEditor) Schema() schema.Schema {
+	return kte.sch
+}
+
+// Name implements TableEditor.
+func (kte *keylessTableEditor) Name() string {
+	return kte.name
+}
+
+// Format implements TableEditor.
+func (kte *keylessTableEditor) Format() *types.NomsBinFormat {
+	return kte.tbl.Format()
+}
+
+// Close implements TableEditor.
+func (kte *keylessTableEditor) Close() error {
+	return nil
+}
+
+// autoFlush will call flush() if we have accumulated enough edits.
+func (kte *keylessTableEditor) autoFlush(ctx context.Context) error {
+	// work is proportional to number of deltas
+	// not the number of row operations
+	if len(kte.acc.deltas) >= tableEditorMaxOps {
+		return kte.flush(ctx)
+	}
+	return nil
+}
+
+func (kte *keylessTableEditor) flush(ctx context.Context) error {
+	// should only wait here on a non-autoFlush
+	// eg kte.Table()
+	err := kte.eg.Wait()
+	if err != nil {
+		return err
+	}
+
+	acc := kte.acc
+	tbl := kte.tbl
+
+	// setup fresh accumulator
+	kte.acc = keylessEditAcc{
+		deltas: make(map[hash.Hash]rowDelta),
+		nbf:    acc.nbf,
+	}
+
+	kte.eg.Go(func() (err error) {
+		kte.tbl, err = applyEdits(ctx, tbl, acc)
+		return err
+	})
+
+	return nil
+}
+
+func applyEdits(ctx context.Context, tbl *doltdb.Table, acc keylessEditAcc) (*doltdb.Table, error) {
+	rowData, err := tbl.GetRowData(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	idx := 0
+	keys := make([]types.Value, len(acc.deltas))
+	for _, vd := range acc.deltas {
+		keys[idx] = vd.key
+		idx++
+	}
+
+	err = types.SortWithErroringLess(types.ValueSort{Values: keys, Nbf: acc.nbf})
+	if err != nil {
+		return nil, err
+	}
+
+	ed := rowData.Edit()
+	iter := table.NewMapPointReader(rowData, keys...)
+
+	for {
+		k, v, err := iter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		delta, err := acc.getRowDelta(k.(types.Tuple))
+		if err != nil {
+			return nil, err
+		}
+
+		var ok bool
+		if v == nil {
+			// row does not yet exist
+			v, ok, err = setCardinality(delta.val, delta.delta)
+		} else {
+			v, ok, err = updateCardinality(v.(types.Tuple), delta.delta)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if ok {
+			ed.Set(k, v)
+		} else {
+			ed.Remove(k)
+		}
+	}
+
+	rowData, err = ed.Map(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return tbl.UpdateRows(ctx, rowData)
+}
+
+// for deletes (cardinality < 1): |ok| is set false
+func setCardinality(val types.Tuple, card int64) (v types.Tuple, ok bool, err error) {
+	if card < 1 {
+		return types.Tuple{}, false, nil
+	}
+
+	v, err = val.Set(cardinalityValIdx, types.Uint(card))
+	if err != nil {
+		return v, false, err
+	}
+
+	return v, true, nil
+}
+
+// for deletes (cardinality < 1): |ok| is set false
+func updateCardinality(val types.Tuple, delta int64) (v types.Tuple, ok bool, err error) {
```

Comment:
set / update are too similar for these two methods, which do very different things. Would call this incrementCardinality or applyDeltaToCardinaolity